# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

## Instructions:

The steps for this challenge are broken out into the following sections:

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model

Using your knowledge of Pandas and scikit-learn’s `StandardScaler()`, preprocess the dataset so that you can use it to compile and evaluate the neural network model later.

Open the starter code file, and complete the following data preparation steps:

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course.

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset.

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use your knowledge of TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset. Consider the number of inputs before determining the number of layers that your model will contain or the number of neurons on each layer. Then, compile and fit your model. Finally, evaluate your binary classification model to calculate the model’s loss and accuracy.

To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

> **Hint** When fitting the model, start with a small number of epochs, such as 20, 50, or 100.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

4. Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`.

### Optimize the Neural Network Model

Using your knowledge of TensorFlow and Keras, optimize your model to improve the model's accuracy. Even if you do not successfully achieve a better accuracy, you'll need to demonstrate at least two attempts to optimize the model. You can include these attempts in your existing notebook. Or, you can make copies of the starter notebook in the same folder, rename them, and code each model optimization in a new notebook.

> **Note** You will not lose points if your model does not achieve a high accuracy, as long as you make at least two attempts to optimize the model.

To do so, complete the following steps:

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. After finishing your models, display the accuracy scores achieved by each model, and compare the results.

3. Save each of your models as an HDF5 file.


In [1]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.


In [2]:
# Import applicants_data.csv to Google Colab
from google.colab import files
uploaded = files.upload()

Saving applicants_data.csv to applicants_data (2).csv


In [3]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
applicant_data_df = pd.read_csv("applicants_data.csv")

# Review the DataFrame
applicant_data_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [4]:
# Review the data types associated with the columns
# YOUR CODE HERE
print(applicant_data_df.dtypes)

EIN                        int64
NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object


### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [5]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame
applicant_data_df = applicant_data_df.drop(["EIN", "NAME"], axis=1)

# Review the DataFrame
# YOUR CODE HERE
applicant_data_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [6]:
# Create a list of categorical variables
categorical_variables = ["APPLICATION_TYPE", "AFFILIATION", "CLASSIFICATION", "USE_CASE", "ORGANIZATION", "INCOME_AMT"]
# Display the categorical variables list
# YOUR CODE HERE
categorical_variables

['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT']

In [7]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse_output=False)

In [8]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])

In [9]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(encoded_data, columns=enc.get_feature_names_out(categorical_variables))
# Review the DataFrame
# YOUR CODE HERE
encoded_df.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  ORGANIZATION_Trust  INCOME_AMT_0  \
0                   0.0  ...                 0.0           1.0   
1                   0.0  ...                 0.0           0.0   
2                   0.0  ...                 0.0           1.0   
3                   0.0  ...                 1.0           0.0   
4                   0.0  ...                 1.0           0.0   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  
0              0.0                0.0  
1              0.0                0.0  
2              0.0                0.0  
3              0.0                0.0  
4              0.0                0.0  

[5 rows x 112 columns]

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course.

In [10]:
# Add the numerical variables from the original DataFrame to the one-hot encoding DataFrame
numerical_variables = ["STATUS", "ASK_AMT", "IS_SUCCESSFUL"]
encoded_df_new = pd.concat([encoded_df, applicant_data_df[numerical_variables]], axis=1)

# Review the DataFrame
encoded_df_new.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_10000-24999  \
0                   0.0  ...                     0.0   
1                   0.0  ...                     0.0   
2                   0.0  ...                     0.0   
3                   0.0  ...                     1.0   
4                   0.0  ...                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  STATUS  \
0                     0.0              0.0                0.0       1   
1                     0.0              0.0                0.0       1   
2                     0.0              0.0                0.0       1   
3                     0.0              0.0                0.0       1   
4                     0.0              0.0                0.0       1   

   ASK_AMT  IS_SUCCESSFUL  
0     5000              1  
1   108590              1  
2     5000              0  
3     6692              1  
4   142590              1  

[5 rows x 115 columns]

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset.



In [11]:
# Define the target set y using the IS_SUCCESSFUL column
y = encoded_df_new["IS_SUCCESSFUL"]
# Display a sample of y
# YOUR CODE HERE
y.head()

0    1
1    1
2    0
3    1
4    1
Name: IS_SUCCESSFUL, dtype: int64

In [12]:
# Define features set X by selecting all columns but IS_SUCCESSFUL
X = encoded_df_new.drop("IS_SUCCESSFUL", axis=1)

# Review the features DataFrame
# YOUR CODE HERE
X.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                   0.0  ...                0.0                     0.0   
1                   0.0  ...                1.0                     0.0   
2                   0.0  ...                0.0                     0.0   
3                   0.0  ...                0.0                     1.0   
4                   0.0  ...                0.0                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  STATUS  ASK_AMT  
0                     0.0              0.0                0.0       1     5000  
1                     0.0              0.0                0.0       1   108590  
2                     0.0              0.0                0.0       1     5000  
3                     0.0              0.0                0.0       1     6692  
4                     0.0              0.0                0.0       1   142590  

[5 rows x 114 columns]

### Step 6: Split the features and target sets into training and testing datasets.


In [13]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [14]:
# Create a StandardScaler instance
scaler = StandardScaler()
# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [15]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train_scaled[0])

# Review the number of features
number_input_features

114

In [16]:
# Define the number of neurons in the output layer
number_output_neurons = 1


In [17]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = 8

# Review the number hidden nodes in the first layer
hidden_nodes_layer1

8

In [18]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = 5

# Review the number hidden nodes in the second layer
hidden_nodes_layer2

5

In [19]:
# Create the Sequential model instance
nn = Sequential()


In [20]:
# Add the first hidden layer
# YOUR CODE HERE
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))


In [21]:
# Add the second hidden layer
# YOUR CODE HERE
nn.add(Dense(units=hidden_nodes_layer2, activation="relu"))


In [22]:
# Add the output layer to the model specifying the number of output neurons and activation function
# YOUR CODE HERE
nn.add(Dense(units=number_output_neurons, activation="sigmoid"))


In [23]:
# Display the Sequential model summary
# YOUR CODE HERE
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 920       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 971
Trainable params: 971
Non-trainable params: 0
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [24]:
# Compile the Sequential model
# YOUR CODE HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [25]:
# Fit the model using 50 epochs and the training data
# YOUR CODE HERE
model = nn.fit(X_train_scaled, y_train, epochs=50)


Epoch 1/50
804/804 [==============================] - 4s 3ms/step - loss: 0.6461 - accuracy: 0.6838
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5680 - accuracy: 0.7237
Epoch 3/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5582 - accuracy: 0.7277
Epoch 4/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5541 - accuracy: 0.7283
Epoch 5/50
804/804 [==============================] - 5s 7ms/step - loss: 0.5521 - accuracy: 0.7300
Epoch 6/50
804/804 [==============================] - 4s 5ms/step - loss: 0.5507 - accuracy: 0.7293
Epoch 7/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5499 - accuracy: 0.7300
Epoch 8/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5486 - accuracy: 0.7310
Epoch 9/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5482 - accuracy: 0.7313
Epoch 10/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5470 - accuracy: 0.7314

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [26]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 [==============================] - 1s 2ms/step - loss: 0.5525 - accuracy: 0.7308
Loss: 0.5524504780769348, Accuracy: 0.7308454513549805


### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`.


In [47]:
# Set the model's file path
file_path = "AlphabetSoup.h5"

# Export your model to a HDF5 file
nn.save(file_path)

# Download your model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

---

## Optimize the neural network model


### Step 1: Define at least three new deep neural network models (resulting in the original plus 3 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.


### Alternative Model 1

In [27]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

114

In [28]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

In [29]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 = 20

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A1

20

In [30]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A1 = 10

# Review the number of hidden nodes in the second layer
hidden_nodes_layer2_A1

10

In [31]:
# Define the number of hidden nodes for the third hidden layer
hidden_nodes_layer3_A1 = 5

# Review the number of hidden nodes in the second layer
hidden_nodes_layer3_A1

5

In [32]:
# Create the Sequential model instance
nn_A1 = Sequential()

In [33]:
# First hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer1_A1, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer2_A1, activation='relu'))

# Third hidden layer
nn_A1.add(Dense(units=hidden_nodes_layer3_A1, activation='relu'))

# Output layer
nn_A1.add(Dense(units=number_output_neurons_A1, activation='sigmoid'))
# Check the structure of the model
# YOUR CODE HERE
print(nn_A1.summary())

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 20)                2300      
                                                                 
 dense_4 (Dense)             (None, 10)                210       
                                                                 
 dense_5 (Dense)             (None, 5)                 55        
                                                                 
 dense_6 (Dense)             (None, 1)                 6         
                                                                 
Total params: 2,571
Trainable params: 2,571
Non-trainable params: 0
_________________________________________________________________
None


In [34]:
# Compile the Sequential model
# YOUR CODE HERE
nn_A1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [35]:
# Fit the model using 50 epochs and the training data
fit_model_A1 = nn_A1.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 2ms/step - loss: 0.5900 - accuracy: 0.7124
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5546 - accuracy: 0.7293
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5511 - accuracy: 0.7295
Epoch 4/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5489 - accuracy: 0.7313
Epoch 5/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5477 - accuracy: 0.7308
Epoch 6/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5472 - accuracy: 0.7308
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5456 - accuracy: 0.7331
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5448 - accuracy: 0.7333
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5435 - accuracy: 0.7331
Epoch 10/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5435 - accuracy: 0.7344

#### Alternative Model 2

In [36]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

114

In [37]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [38]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A2 = 50

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A2

50

In [39]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A2 = 25

# Review the number of hidden nodes in the second layer
hidden_nodes_layer2_A2

25

In [40]:
# Create the Sequential model instance
nn_A2 = Sequential()

In [41]:
# First hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer1_A2, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer2_A2, activation='relu'))

# Output layer
nn_A2.add(Dense(units=number_output_neurons_A2, activation='sigmoid'))
# Check the structure of the model
# YOUR CODE HERE
print(nn_A2.summary())

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 50)                5750      
                                                                 
 dense_8 (Dense)             (None, 25)                1275      
                                                                 
 dense_9 (Dense)             (None, 1)                 26        
                                                                 
Total params: 7,051
Trainable params: 7,051
Non-trainable params: 0
_________________________________________________________________
None


In [42]:
# Compile the model
# YOUR CODE HERE
nn_A2.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [43]:
# Fit the model
fit_model_A2 = nn_A2.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 4s 3ms/step - loss: 0.5755 - accuracy: 0.7209
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5533 - accuracy: 0.7282
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5489 - accuracy: 0.7298
Epoch 4/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5471 - accuracy: 0.7318
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5454 - accuracy: 0.7327
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5442 - accuracy: 0.7323
Epoch 7/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5434 - accuracy: 0.7342
Epoch 8/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5430 - accuracy: 0.7336
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5421 - accuracy: 0.7334
Epoch 10/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5424 - accuracy: 0.7350

### Step 2: After finishing your models, display the accuracy scores achieved by each model, and compare the results.

In [44]:
print("Original Model Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test)
# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Original Model Results
268/268 [==============================] - 0s 2ms/step - loss: 0.5525 - accuracy: 0.7308
Loss: 0.5524504780769348, Accuracy: 0.7308454513549805


In [45]:
print("Alternative Model 1 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn_A1.evaluate(X_test_scaled, y_test)
# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 1 Results
268/268 [==============================] - 1s 2ms/step - loss: 0.5512 - accuracy: 0.7307
Loss: 0.5512396097183228, Accuracy: 0.7307288646697998


In [46]:
print("Alternative Model 2 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn_A2.evaluate(X_test_scaled, y_test)
# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 2 Results
268/268 [==============================] - 1s 3ms/step - loss: 0.5539 - accuracy: 0.7291
Loss: 0.5539257526397705, Accuracy: 0.7290962338447571


### Step 3: Save each of your alternative models as an HDF5 file.


In [48]:
# Set the file path for the first alternative model
file_path = "AlphabetSoup_A1.h5"

# Export your model to a HDF5 file
nn_A1.save(file_path)

# Download the model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [49]:
# Set the file path for the second alternative model
file_path = "AlphabetSoup_A2.h5"

# Export your model to a HDF5 file
nn_A2.save(file_path)

# Download the model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Personal Experiment

## Alternative Model A3

I noticed my models accuracy declined across the 3 variations.

This is an experiment to see if I can create one more alternative model that exhibits an increased accuracy.

In [58]:
# Define the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

In [59]:
# Review the number of features
print(number_input_features)

114


In [60]:
# Define the number of neurons in the output layer
number_output_neurons_A3 = 1

In [61]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A3 = 100

In [62]:
# Review the number of hidden nodes in the first layer
print(hidden_nodes_layer1_A3)

100


In [63]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A3 = 50

In [64]:
# Review the number of hidden nodes in the second layer
print(hidden_nodes_layer2_A3)

50


In [65]:
# Create the Sequential model instance
nn_A3 = Sequential()

In [66]:
# First hidden layer
nn_A3.add(Dense(units=hidden_nodes_layer1_A3, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn_A3.add(Dense(units=hidden_nodes_layer2_A3, activation='relu'))

# Output layer
nn_A3.add(Dense(units=number_output_neurons_A3, activation='sigmoid'))

In [67]:
# Check the structure of the model
print(nn_A3.summary())

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_10 (Dense)            (None, 100)               11500     
                                                                 
 dense_11 (Dense)            (None, 50)                5050      
                                                                 
 dense_12 (Dense)            (None, 1)                 51        
                                                                 
Total params: 16,601
Trainable params: 16,601
Non-trainable params: 0
_________________________________________________________________
None


In [68]:
# Compile the model
nn_A3.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [69]:
# Fit the model
fit_model_A3 = nn_A3.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 4s 3ms/step - loss: 0.5747 - accuracy: 0.7200
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5533 - accuracy: 0.7290
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5495 - accuracy: 0.7303
Epoch 4/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5472 - accuracy: 0.7321
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5456 - accuracy: 0.7340
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5447 - accuracy: 0.7345
Epoch 7/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5437 - accuracy: 0.7342
Epoch 8/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5431 - accuracy: 0.7328
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5425 - accuracy: 0.7337
Epoch 10/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5417 - accuracy: 0.7363

In [70]:
print("Alternative Model 3 Results")
model_loss, model_accuracy = nn_A3.evaluate(X_test_scaled, y_test)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 3 Results
268/268 [==============================] - 1s 2ms/step - loss: 0.5541 - accuracy: 0.7304
Loss: 0.5540934205055237, Accuracy: 0.7303789854049683


In [71]:
# Set the file path for the second alternative model
file_path = "AlphabetSoup_A3.h5"

# Export your model to a HDF5 file
nn_A3.save(file_path)

# Download the model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [72]:
# Display all accuracy scores
print("Original Model Results")
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

print("Alternative Model 1 Results")
model_loss, model_accuracy = nn_A1.evaluate(X_test_scaled, y_test)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

print("Alternative Model 2 Results")
model_loss, model_accuracy = nn_A2.evaluate(X_test_scaled, y_test)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

print("Alternative Model 3 Results")
model_loss, model_accuracy = nn_A3.evaluate(X_test_scaled, y_test)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


Original Model Results
268/268 [==============================] - 1s 3ms/step - loss: 0.5525 - accuracy: 0.7308
Loss: 0.5524504780769348, Accuracy: 0.7308454513549805
Alternative Model 1 Results
268/268 [==============================] - 1s 3ms/step - loss: 0.5512 - accuracy: 0.7307
Loss: 0.5512396097183228, Accuracy: 0.7307288646697998
Alternative Model 2 Results
268/268 [==============================] - 1s 2ms/step - loss: 0.5539 - accuracy: 0.7291
Loss: 0.5539257526397705, Accuracy: 0.7290962338447571
Alternative Model 3 Results
268/268 [==============================] - 0s 2ms/step - loss: 0.5541 - accuracy: 0.7304
Loss: 0.5540934205055237, Accuracy: 0.7303789854049683


# Summary:

In this activity, we built and trained four different models to predict the success of a charitable organization's funding using TensorFlow and Keras. The models were trained on a dataset containing various features such as application type, organization type, and income classification.

The original model achieved an accuracy of 73.08%, while the alternative models showed slightly lower accuracies: Alternative Model 1 achieved 73.07% accuracy, Alternative Model 2 achieved 72.91% accuracy, and Alternative Model 3 achieved 73.04% accuracy.

Although Alternative Model 3 did not significantly improve the accuracy compared to the original model, it offers a different configuration of hidden layers and nodes. Further experimentation and adjustments to the model's architecture may be necessary to achieve higher accuracy.

All models were saved as HDF5 files for future use.